bindings/nodejs/lib/types/output.ts

```diff
@@ -0,0 +1,241 @@
+     */
+    getUnlockConditions(): UnlockCondition[] {
+        return this.unlockConditions;
+    }
+    /**
+     * The native tokens held by the output.
+     */
+    getNativeTokens(): INativeToken[] | undefined {
+        return this.nativeTokens;
+    }
+
+    setNativeTokens(tokens: INativeToken[] | undefined) {
+        this.nativeTokens = tokens;
+    }
+    /**
+     * Features contained by the output.
+     */
+    getFeatures(): Feature[] | undefined {
+        return this.features;
+    }
+
+    setFeatures(features: Feature[] | undefined) {
+        this.features = features;
+    }
+}
+/**
+ * Treasury Output.
+ */
+class TreasuryOutput extends Output /*implements ITreasuryOutput */ {
+    constructor(amount: string) {
+        super(OutputType.Treasury, amount);
+    }
+}
+/**
+ * Basic output.
+ */
+class BasicOutput extends CommonOutput /*implements IBasicOutput*/ {
+    constructor(amount: string, unlockConditions: UnlockCondition[]) {
+        super(OutputType.Basic, amount, unlockConditions);
+    }
+}
+
+abstract class ImmutableFeaturesOutput extends CommonOutput {
+    private immutableFeatures?: Feature[];
+
+    constructor(
+        type: OutputType,
+        amount: string,
+        unlockConditions: UnlockCondition[],
+    ) {
+        super(type, amount, unlockConditions);
+    }
+    /**
+     * Immutable features contained by the output.
+     */
+    getImmutableFeatures(): Feature[] | undefined {
+        return this.immutableFeatures;
+    }
+
+    setImmutableFeatures(immutableFeatures: Feature[] | undefined) {
+        this.immutableFeatures = immutableFeatures;
+    }
+}
+
+abstract class StateMetadataOutput extends ImmutableFeaturesOutput /*implements IBasicOutput*/ {
+    private stateMetadata?: HexEncodedString;
+
+    constructor(
+        type: OutputType,
+        amount: string,
+        unlockConditions: UnlockCondition[],
+    ) {
+        super(type, amount, unlockConditions);
+    }
+    /**
+     * Metadata that can only be changed by the state controller.
+     */
+    getStateMetadata(): HexEncodedString | undefined {
+        return this.stateMetadata;
+    }
+
+    setStateMetadata(stateMetadata: HexEncodedString | undefined) {
+        this.stateMetadata = stateMetadata;
+    }
+}
+
+class AliasOutput extends StateMetadataOutput /*implements IAliasOutput*/ {
+    private aliasId: HexEncodedString;
+    private stateIndex: number;
+    private foundryCounter: number;
+
+    constructor(
+        unlockConditions: UnlockCondition[],
+        amount: string,
+        aliasId: HexEncodedString,
+        stateIndex: number,
+        foundryCounter: number,
+    ) {
+        super(OutputType.Alias, amount, unlockConditions);
+        this.aliasId = aliasId;
+        this.stateIndex = stateIndex;
+        this.foundryCounter = foundryCounter;
+    }
+    /**
+     * Unique identifier of the alias, which is the BLAKE2b-160 hash of the Output ID that created it.
```

Comment:
For the first output it's null, do we want to mention that?

Reply:
We could mention, yea. where is this used/described? Want to read up on it

Reply:
https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md#additional-transaction-semantic-validation-rules-1
search for `zeroed out`